### PR TITLE
ROL documentation

### PR DESCRIPTION
**Pull request by APaganini**
This branch introduces some documentation about using ROL in fireshape.

### CODE REVIEW

File: docs/source/ROL.rst [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    constraint. Among other reasons, this can happen when the control
+    :bash:`q` is not feasible (for instance, when the underlying mesh interstects itself)
+    or when the state constraint is nonlinear and the optimization step
+    is too large (in which case the initial guess is not good enough).
+
+.. note::
+
+    The following examples include all parameters that can be set
+    for the algorithms described. However, it is not necessary to
+    specify a field if one does want to modify a default value.
```
Comment (florianwechsung) [inline docs/source/ROL.rst]: do you mean "if one does not"?

File: docs/source/ROL.rst [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+.. note::
+
+    The following examples include all parameters that can be set
+    for the algorithms described. However, it is not necessary to
+    specify a field if one does want to modify a default value.
```
Comment (florianwechsung) [inline docs/source/ROL.rst]: This note seems to be repeated?

File: docs/source/ROL.rst [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+Solving a constrained problem with an augmented Lagrangian method
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To solve a problem with equality or inequality constraint
+using an augmented Lagrangian method,
+we can set :bash:`Step` and :bash:`General` as follows
+(the provided values are the default ones).
+Note that the augmented Lagrangian algorithm solves
+a sequence of surrogate models. These surrogate models
+are unconstrained optimization probelms that encode constraints
```
Comment (florianwechsung) [inline docs/source/ROL.rst]: probelms -> problems

File: docs/source/ROL.rst [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+Solving a constrained problem with an augmented Lagrangian method
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To solve a problem with equality or inequality constraint
+using an augmented Lagrangian method,
+we can set :bash:`Step` and :bash:`General` as follows
+(the provided values are the default ones).
+Note that the augmented Lagrangian algorithm solves
+a sequence of surrogate models. These surrogate models
```
Comment (florianwechsung) [inline docs/source/ROL.rst]: The term "surrogate model" has a very specific meaning in the optimisation case and I don't think is appropriate here. I suggest "auxiliary" or "intermediate" problems.

File: docs/source/ROL.rst [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+.. note::
+
+    We prefer using a trust-region method instead of a line-search
+    method because in the former case it is easier to deal with
+    situations when the routine that evaluates the functional :bash:`J`
+    fails. Such failures are usually due to failure in solving the state
+    constraint. Among other reasons, this can happen when the control
+    :bash:`q` is not feasible (for instance, when the underlying mesh interstects itself)
+    or when the state constraint is nonlinear and the optimization step
+    is too large (in which case the initial guess is not good enough).
```
Comment (florianwechsung) [inline docs/source/ROL.rst]: Can you elaborate why a trust region algorithm struggles less when we can't evaluate the objective?

File: docs/source/ROL.rst [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    The variables :bash:`emul` and :bash:`imul` are of type :bash:`ROL.StdVector`.
+    The variable :bash:`ibnd` is of type :bash:`ROL.BoundConstraint`. For example,
+    the following code sets the lower and the upper bound to 2 and 7, respectively.
+
+    .. code-block::
+
+        lower = ROL.StdVector(1); lower[0] = 2
+        upper = ROL.StdVector(1); upper[0] = 7
+        ibnd = ROL.BoundConstraint(lower, upper)
+        ??shall we also include flags to activate??
```
Comment (florianwechsung) [inline docs/source/ROL.rst]: looking at the constructors in https://bitbucket.org/pyrol/trilinos/src/pyrol/packages/rol/src/function/boundconstraint/ROL_Bounds.hpp, I don't think we have to worry about the `Activate` class. Or is there anything in particular in the code that you found that suggests we need to deal with it?

File: docs/source/ROL.rst (hunk 1 above)
Comment (APaganini) [reply docs/source/ROL.rst]: yes, thanks

File: docs/source/ROL.rst (hunk 2 above)
Comment (APaganini) [reply docs/source/ROL.rst]: removed, thanks

File: docs/source/ROL.rst (hunk 3 above)
Comment (APaganini) [reply docs/source/ROL.rst]: corrected

File: docs/source/ROL.rst (hunk 4 above)
Comment (APaganini) [reply docs/source/ROL.rst]: good point, I use intermediate

File: docs/source/ROL.rst (hunk 5 above)
Comment (APaganini) [reply docs/source/ROL.rst]: I've added more details about trust-region methods

File: docs/source/ROL.rst (hunk 6 above)
Comment (APaganini) [reply docs/source/ROL.rst]: We don't need anything special. I just wanted to confirm that we don't need it. Thanks for confirming

File: docs/source/ROL.rst [hunk 7]
```diff
@@ -105,21 +104,42 @@ method.
...
-    or when the state constraint is nonlinear and the optimization step
-    is too large (in which case the initial guess is not good enough).
+    :bash:`q` is not feasible (for instance, when the underlying mesh
+    interstects itself) or when the state constraint is nonlinear and
+    the optimization step is too large (in which case the initial guess
+    is not good enough).
+
+    In a nutshell, trust-region methods solve a sequence of optimization
+    problems. In each of these, one minimizes a quadratic misfit functional
+    with control constraints. The idea is that the quadratic misfict functional
```
Comment (florianwechsung) [inline docs/source/ROL.rst]: typo: misfict

File: docs/source/ROL.rst [hunk 8]
```diff
@@ -105,21 +104,42 @@ method.
...
-    :bash:`q` is not feasible (for instance, when the underlying mesh interstects itself)
-    or when the state constraint is nonlinear and the optimization step
-    is too large (in which case the initial guess is not good enough).
+    :bash:`q` is not feasible (for instance, when the underlying mesh
+    interstects itself) or when the state constraint is nonlinear and
+    the optimization step is too large (in which case the initial guess
+    is not good enough).
+
+    In a nutshell, trust-region methods solve a sequence of optimization
+    problems. In each of these, one minimizes a quadratic misfit functional
```
Comment (florianwechsung) [inline docs/source/ROL.rst]: I don't see where the _misfit_ functional comes from. Do you just mean a quadratic approximation to the true objective? I associate the term _misfit_ with something like a l^2 minimization, which we don't necessarily do.

File: docs/source/ROL.rst [hunk 9]
```diff
@@ -105,21 +104,42 @@ method.
...
+    Then, if there is good agreement between the actual and predicted
+    reductions, the trust-region radius is increased. This radius is
+    decreased if the actual reduction is not positive or the ratio between
+    actual and predicted reductions is close to zero.
+    From this, we undestand that a safe solution to deal with failed
+    evaluations of :bash:`J` is to store the previously computed value of
+    :bash:`J` and, using a :bash:`try: ... except: ...` approach,
+    return it if the new evaluation of :bash:`J` fails. This corresponds
+    to a nonpositive actual reduction, which triggers a reduction of
+    the trust-region radius.
```
Comment (florianwechsung) [inline docs/source/ROL.rst:136]: Do we need to actually store the value or could we just return infinity? The latter would simply the code quite a bit

File: docs/source/ROL.rst [hunk 10]
```diff
@@ -105,21 +104,42 @@ method.
     situations when the routine that evaluates the functional :bash:`J`
     fails. Such failures are usually due to failure in solving the state
     constraint. Among other reasons, this can happen when the control
-    :bash:`q` is not feasible (for instance, when the underlying mesh interstects itself)
-    or when the state constraint is nonlinear and the optimization step
-    is too large (in which case the initial guess is not good enough).
+    :bash:`q` is not feasible (for instance, when the underlying mesh
+    interstects itself) or when the state constraint is nonlinear and
```
Comment (florianwechsung) [inline docs/source/ROL.rst]: typo: interstects

File: docs/source/ROL.rst [hunk 11]
```diff
@@ -105,21 +104,42 @@ method.
...
     The following examples include all parameters that can be set
     for the algorithms described. However, it is not necessary to
-    specify a field if one does want to modify a default value.
-
-.. note::
-
-    The following examples include all parameters that can be set
-    for the algorithms described. However, it is not necessary to
-    specify a field if one does want to modify a default value.
+    specify a field if one does want not to modify a default value.
```
Comment (florianwechsung) [inline docs/source/ROL.rst]: one does want not -> one does not want to

File: docs/source/ROL.rst (hunk 7 above)
Comment (APaganini) [reply docs/source/ROL.rst]: removed misfict

File: docs/source/ROL.rst (hunk 8 above)
Comment (APaganini) [reply docs/source/ROL.rst]: I dropped misfict and, when necessary, replaced it with objective

File: docs/source/ROL.rst (hunk 10 above)
Comment (APaganini) [reply docs/source/ROL.rst]: fixed

File: docs/source/ROL.rst (hunk 11 above)
Comment (APaganini) [reply docs/source/ROL.rst]: fixed

**Review by florianwechsung — APPROVED** [review]

**Pull request merged**